pkg/controller/elasticsearch/sset/validation.go

```diff
@@ -0,0 +1,90 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package sset
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/rand"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type PodTemplateError struct {
+	Parent      metav1.Object
+	StatefulSet appsv1.StatefulSet
+	Causes      []metav1.StatusCause
+}
+
+func (e *PodTemplateError) Error() string {
+	return fmt.Sprintf(
+		"Validation of PodTemplate for StatefulSet %s in Elasticsearch %s/%s failed for the following reasons: %v",
+		e.StatefulSet.Name,
+		e.Parent.GetNamespace(),
+		e.Parent.GetName(),
+		e.Causes,
+	)
+}
+
+// validatePodTemplate validates a Pod Template by issuing a dry API request.
+func validatePodTemplate(
+	c k8s.Client,
+	parent metav1.Object,
+	sset appsv1.StatefulSet,
+) error {
+	template := sset.Spec.Template
+	// Create a dummy Pod with the pod template
+	dummyPod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:   sset.GetNamespace(),
+			Name:        sset.GetName() + "-dummy-" + rand.String(5),
+			Labels:      template.Labels,
+			Annotations: template.Annotations,
+		},
+		Spec: template.Spec,
+	}
+	if err := c.Create(dummyPod, client.DryRunAll); err != nil {
+		return toPodTemplateError(parent, sset, err)
+	}
+	return nil
+}
+
+// toPodTemplateError attempts to extract the meaningful information from the dry run API call by converting the original
+// error into a PodTemplateError.
+func toPodTemplateError(
+	parent metav1.Object,
+	sset appsv1.StatefulSet,
+	err error,
+) error {
+	var statusErr *k8serrors.StatusError
+	if !errors.As(err, &statusErr) {
+		// Not a Kubernetes API error (e.g. timeout)
```

Comment:
I'm wondering if we could always reach a timeout in some situations (eg. particular proxy configuration). Hard to know for sure, so I'm fine moving on with this.
Should we add some debug logs here just in case?

Reply:
I think that if we reach a timeout here it is either:
* a transient error, in this case it is fair to try again in the next reconciliation loop
* a permanent error, in which case the StatefulSet update/creation that should happen right after will also fail

The error is already reported in the controller log, I'm not sure we should duplicate it.